### PR TITLE
fix:ethapi:Fail fast if user attempts to send to 0x000..

### PR DESCRIPTION
**Pull request by ZenGround0**
## Proposed Changes

This is broken silently now and its not good for UX

## Additional Info

https://filecoinproject.slack.com/archives/CRK2LKYHW/p1717635018083559

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [x] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by ZenGround0** [comment]
It would be better to actually make this work.  For now I'm making the implicit error (attempt to run invoke contract and get rekt) explicit.  Now users will see rpc server error instead of a weird version of success.

To actually make this work we'd need to mix up things in a difficult way.  There is no longer one canonical serialization of filecoin create external so we'd need to test out both against sig and go with the one that validates.

**Review by rvagg — APPROVED** [review]

**Comment by ZenGround0** [comment]
Ok nice this fails explicitly now:
```
Error: 
server returned an error response: error code 1: failed to convert to unsigned msg: failed to get method info: zero address recipient disallowed, please set `To` to null
```
I'm going to try to get it working to handle the 0x000 case too.

File: chain/types/ethtypes/eth_transactions.go [hunk 1]
```diff
@@ -253,6 +253,10 @@ func getFilecoinMethodInfo(recipient *EthAddress, input []byte) (*methodInfo, er
 	var to address.Address
 	var method abi.MethodNum
 
+	var zeroEthAddr EthAddress
+	if recipient != nil && *recipient == zeroEthAddr {
+		return nil, fmt.Errorf("zero address recipient disallowed, please set `To` to null")
+	}
```
Comment (Stebalien) [inline chain/types/ethtypes/eth_transactions.go:259]: This is used inside message validation (now) and is therefore consensus critical. Given that the user _can_ send to this address (it is a valid account), we're not allowed to do this.

**Review by Stebalien — CHANGES_REQUESTED** [review]
Tooling _shouldn't_ send to 0x0 when creating contracts. if you're seeing that, something is very wrong somewhere else in the API.

**Comment by Stebalien** [comment]
Yeah, I've confirmed:

1. Sends to 0x0 are _valid_.
2. Contract creation transactions send to RLP-encoded `nil` (0x80).

If we're somehow translating this `nil` to `0x0000000...`, we have a bigger problem.

**Comment by ZenGround0** [comment]
I'll keep trying to understand where the 0x0 is coming from.  Hopefully just the tool.

File: chain/types/ethtypes/eth_transactions.go (hunk 1 above)
Comment (ZenGround0) [reply chain/types/ethtypes/eth_transactions.go:259]: FYI @rvagg this was a pretty bad mistake on my part.  If I'd merged this and we hadn't thought more about it we would have opened up a fork.  So I guess the lesson for you here is to trust me less.

**Pull request closed without merging**